crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+                )
+                .and_then(|m| {
+                    if page_index < m.size() as usize {
+                        Some(page_index)
+                    } else {
+                        None
+                    }
+                });
+
+            return Some(AddressLocation::MemoryPage {
+                page_addr: (memory_start + page_index * WASM_PAGE_SIZE) as _,
+                len: WASM_PAGE_SIZE,
+                instance,
+                memory_index,
+                page_index: init_page_index,
+            });
+        }
+
+        // Check for a table location
+        if addr >= self.tables_start && addr < self.tables_end {
+            let index = (addr - self.tables_start) / self.table_size;
+            let table_start = self.tables_start + (index * self.table_size);
+            let table_offset = addr - table_start;
+            let page_index = table_offset / self.page_size;
+
+            return Some(AddressLocation::TablePage {
+                page_addr: (table_start + (page_index * self.page_size)) as _,
+                len: self.page_size,
+            });
+        }
+
+        // Check for a stack location
+        if addr >= self.stacks_start && addr < self.stacks_end {
+            let index = (addr - self.stacks_start) / self.stack_size;
+            let stack_start = self.stacks_start + (index * self.stack_size);
+            let stack_offset = addr - stack_start;
+            let page_offset = (stack_offset / self.page_size) * self.page_size;
+
+            return Some(AddressLocation::StackPage {
+                page_addr: (stack_start + page_offset) as _,
+                len: self.page_size,
+                index,
+                guard_page: stack_offset < self.page_size,
+            });
+        }
+
+        None
+    }
+}
+
+unsafe fn wake_guard_page_access(
+    uffd: &Uffd,
+    page_addr: *const u8,
+    len: usize,
+) -> Result<(), String> {
+    // Set the page to NONE to induce a SIGSEV for the access on the next retry
+    region::protect(page_addr, len, region::Protection::NONE)
+        .map_err(|e| format!("failed to change guard page protection: {}", e))?;
+
+    uffd.wake(page_addr as _, len).map_err(|e| {
+        format!(
+            "failed to wake page at {:p} with length {}: {}",
+            page_addr, len, e
+        )
+    })?;
+
+    Ok(())
+}
+
+unsafe fn initialize_wasm_page(
+    uffd: &Uffd,
+    instance: &Instance,
+    page_addr: *const u8,
+    memory_index: usize,
+    page_index: usize,
+) -> Result<(), String> {
+    if let Some(MemoryInitialization::Paged { page_size, map }) =
+        &instance.module.memory_initialization
+    {
+        let memory_index = DefinedMemoryIndex::new(memory_index);
+        let memory = instance.memory(memory_index);
+        let pages = &map[memory_index];
+        debug_assert_eq!(WASM_PAGE_SIZE % page_size, 0);
+
+        let count = WASM_PAGE_SIZE / page_size;
+        let start = page_index * count;
```

Comment:
This is one part where the native page size and wasm page size sent me spinning for a bit. It looks like the inputs to this function are wasm page size indices and multiples, but the `map` that we're loading from is stored with native page sizes aligned. How come that's the case? Should the `map` have wasm page sizes instead? (or should this function take native page size multiples/indices?)

Reply:
The reason the map stores OS pages while the uffd initialization is Wasm-page sized is that I was concerned over how much we might potentially allocate as part of module compilation to represent potentially sparse and disjointed data segments.

I think if we changed the implementation such that we only ever use the "paged" initialization when on Linux with uffd enabled, then I think we should change the map to be Wasm-page sized and give the kernel a single call to copy the page when handling the fault here.